Build Docker image for ARM, ARM64 (M1)

Is it possible to build a Docker image for ARM and ARM64 to run on Raspberry Pi?

I tried building a container on my raspberry pi but ran into an issue with a missing main.js file:

```
Sending build context to Docker daemon  84.75MB
Step 1/12 : FROM alpine:3.12
3.12: Pulling from library/alpine
420c7481a3a7: Pull complete
Digest: sha256:36553b10a4947067b9fbb7d532951066293a68eae893beba1d9235f7d11a20ad
Status: Downloaded newer image for alpine:3.12
 ---> 36a8b93b867d
Step 2/12 : ENV PORT 8080
 ---> Running in aa97abef4489
Removing intermediate container aa97abef4489
 ---> f992bf32726a
Step 3/12 : ENV NODE_ENV=dev
 ---> Running in da3eb2bd8b4b
Removing intermediate container da3eb2bd8b4b
 ---> d36c149addff
Step 4/12 : ENV NC_VERSION=0.6
 ---> Running in 4df2cbb8741c
Removing intermediate container 4df2cbb8741c
 ---> b15dedee6cc1
Step 5/12 : ENV NC_DOCKER=0.6
 ---> Running in a2a1d9764f89
Removing intermediate container a2a1d9764f89
 ---> 54aff99fb420
Step 6/12 : WORKDIR /usr/src/appTemp
 ---> Running in 1246787cadfd
Removing intermediate container 1246787cadfd
 ---> 35519430b3fd
Step 7/12 : COPY ./docker/main.js ./docker/main.js
COPY failed: file not found in build context or excluded by .dockerignore: stat
docker/main.js: file does not exist
```

It seems to be missing because its in the .gitignore located here:

 nocodb/packages/nocodb/.gitignore

I tried building with that line commented out of the dockerfile but that led to this error once the container was running:

```

Error: Cannot find module '/usr/src/app/docker/main.js'


    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:815:15)


    at Function.Module._load (internal/modules/cjs/loader.js:667:27)


    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12)


    at internal/main/run_main_module.js:17:47 {


  code: 'MODULE_NOT_FOUND',


  requireStack: []


}


internal/modules/cjs/loader.js:818
```

> I tried building a container on my raspberry pi but ran into an issue with a missing main.js file:
> 
> ```
> Sending build context to Docker daemon  84.75MB
> Step 1/12 : FROM alpine:3.12
> 3.12: Pulling from library/alpine
> 420c7481a3a7: Pull complete
> Digest: sha256:36553b10a4947067b9fbb7d532951066293a68eae893beba1d9235f7d11a20ad
> Status: Downloaded newer image for alpine:3.12
>  ---> 36a8b93b867d
> Step 2/12 : ENV PORT 8080
>  ---> Running in aa97abef4489
> Removing intermediate container aa97abef4489
>  ---> f992bf32726a
> Step 3/12 : ENV NODE_ENV=dev
>  ---> Running in da3eb2bd8b4b
> Removing intermediate container da3eb2bd8b4b
>  ---> d36c149addff
> Step 4/12 : ENV NC_VERSION=0.6
>  ---> Running in 4df2cbb8741c
> Removing intermediate container 4df2cbb8741c
>  ---> b15dedee6cc1
> Step 5/12 : ENV NC_DOCKER=0.6
>  ---> Running in a2a1d9764f89
> Removing intermediate container a2a1d9764f89
>  ---> 54aff99fb420
> Step 6/12 : WORKDIR /usr/src/appTemp
>  ---> Running in 1246787cadfd
> Removing intermediate container 1246787cadfd
>  ---> 35519430b3fd
> Step 7/12 : COPY ./docker/main.js ./docker/main.js
> COPY failed: file not found in build context or excluded by .dockerignore: stat
> docker/main.js: file does not exist
> ```
> 
> It seems to be missing because its in the .gitignore located here:
> 
> nocodb/packages/nocodb/.gitignore
> 
> I tried building with that line commented out of the dockerfile but that led to this error once the container was running:
> 
> ```
> 
> Error: Cannot find module '/usr/src/app/docker/main.js'
> 
> 
>     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:815:15)
> 
> 
>     at Function.Module._load (internal/modules/cjs/loader.js:667:27)
> 
> 
>     at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12)
> 
> 
>     at internal/main/run_main_module.js:17:47 {
> 
> 
>   code: 'MODULE_NOT_FOUND',
> 
> 
>   requireStack: []
> 
> 
> }
> 
> 
> internal/modules/cjs/loader.js:818
> ```

Generate `main.js` file using following commands

```
npm run build
npm run docker:build
```

--- a/.github/workflows/docker.yml
+++ b/.github/workflows/docker.yml
@@ -1,0 +1,77 @@
+name: docker
+
+on:
+  workflow_dispatch:
+    inputs:
+      tag:
+        description: "Docker image tag"
+        required: true
+
+jobs:
+  buildx:
+    runs-on: ubuntu-latest
+    env:
+      working-directory: ./packages/nocodb
+    strategy:
+      matrix:
+        node-version: [12]
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Use Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@v2
+        with:
+          node-version: ${{ matrix.node-version }}
+
+      - uses: bahmutov/npm-install@v1
+        with:
+          working-directory: ${{ env.working-directory }}
+
+      - name: Build nocodb and docker files
+        run: |
+          npm run build
+          npm run docker:build
+        working-directory: ${{ env.working-directory }}
+
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
+
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Cache Docker layers
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-buildx-
+
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      - name: Build and push
+        uses: docker/build-push-action@v2
+        with:
+          context: ${{ env.working-directory }}
+          build-args: NC_VERSION=${{ github.event.inputs.tag }}
+          platforms: linux/amd64,linux/arm64,linux/arm/v7
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache-new
+          push: true
+          tags: |
+            nocodb/nocodb:${{ github.event.inputs.tag }}
+            nocodb/nocodb:latest
+
+        # Temp fix
+        # https://github.com/docker/build-push-action/issues/252
+        # https://github.com/moby/buildkit/issues/1896
+      - name: Move cache
+        run: |
+          rm -rf /tmp/.buildx-cache
+          mv /tmp/.buildx-cache-new /tmp/.buildx-cache

--- a/packages/nocodb/Dockerfile
+++ b/packages/nocodb/Dockerfile
@@ -1,52 +1,44 @@
-FROM alpine:3.12
+###########
+# Builder
+###########
+FROM node:12 as builder
+WORKDIR /usr/src/app
 
-# Bug fix for segfault ( Convert PT_GNU_STACK program header into PT_PAX_FLAGS )
-#RUN apk --update --no-cache add paxctl \
-#	&& paxctl -cm $(which node)
-
+ENV NODE_ENV production
+ENV NC_VERSION 0.6
+ENV NC_DOCKER 0.6
 ENV PORT 8080
-ENV NODE_ENV=dev
-#ENV TOOL_DIR=/tool
-ENV NC_VERSION=0.6
-ENV NC_DOCKER=0.6
-
-# Create and change to the app directory.
-WORKDIR /usr/src/appTemp
 
 # Copy application dependency manifests to the container image.
 # A wildcard is used to ensure both package.json AND package-lock.json are copied.
-# Copying this separately prevents re-running npm install on every code change.
-#COPY ./build/ ./build/
+# Copying this separately prevents re-running npm ci on every code change.
+COPY ./package*.json ./
 COPY ./docker/main.js ./docker/main.js
-COPY ./package.json ./
 COPY ./docker/start.sh /usr/src/appEntry/start.sh
 
-# Install production dependencies.
-#RUN npm i ../xc-lib-gui/ && npm install --cache=/usr/src/app/cache --production && npx modclean  --patterns="default:*" --run && rm -rf /usr/src/app/cache && rm -rf /root/.npm
-RUN apk --update --no-cache add \
-    	nodejs \
-    	nodejs-npm \
-    	tar\
-   && npm install --cache=/usr/src/app/cache --production \
+# install production dependencies,
+# reduce node_module size with modclean & removing sqlite deps,
+# package built code into app.tar.gz & add execute permission to start.sh
+RUN npm ci --production --quiet \
     && npx modclean --patterns="default:*" --ignore="nc-lib-gui/**,dayjs/**,express-status-monitor/**" --run  \
     && rm -rf ./node_modules/sqlite3/deps \
-#    && rm -rf ./node_modules/nc-lib-gui/lib/dist/_nuxt \
-    && rm -rf /usr/src/app/cache && rm -rf /root/.npm \
-   && apk del nodejs-npm \
-   && tar -czf ../appEntry/app.tar.gz ./* ; rm -rf ./* && chmod +x /usr/src/appEntry/start.sh
+    && tar -czf ../appEntry/app.tar.gz ./* \
+    && chmod +x /usr/src/appEntry/start.sh
 
-
-#COPY ./node_modules/xc-lib/ ./node_modules/xc-lib/
-#COPY ./node_modules/xc-lib-gui/ ./node_modules/xc-lib-gui/
+##########
+# Runner
+##########
+FROM alpine:3.12
 WORKDIR /usr/src/app
 
-# Run the web service on container startup.
-#CMD [ "node", "docker/index.js" ]
+RUN apk --update --no-cache add \
+    nodejs \
+    tar
+
+# Copy packaged production code & main entry file
+COPY --from=builder /usr/src/appEntry/ /usr/src/appEntry/
+
+EXPOSE 8080
+
+# Start Nocodb
 ENTRYPOINT ["sh", "/usr/src/appEntry/start.sh"]
-
-
-
-
-
-
-

